src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,745 @@
+	}
+}
+
+static void crossover_reset_state(struct comp_data *cd)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_ch(&cd->state[i]);
+}
+
+/**
+ * The assign_sink array in the configuration maps to the pipeline ids.
+ * This function  returns the index i such that assign_sink[i] = pipe_id
+ */
+static uint8_t get_stream_index(struct sof_crossover_config *config,
+				uint32_t pipe_id)
+{
+	int i;
+	int32_t *assign_sink = &config->data[0];
+
+	for (i = 0; i < config->num_sinks; i++)
+		if (assign_sink[i] == pipe_id)
+			return i;
+
+	comp_cl_err(&comp_crossover, "get_stream_index() error: couldn't find configuration for connected pipeline %u",
+		    pipe_id);
+
+	return -EINVAL;
+}
+
+static int crossover_assign_sinks(struct comp_dev *dev,
+				  struct sof_crossover_config *config,
+				  struct comp_buffer **sinks)
+{
+	struct comp_buffer *sink;
+	struct list_item *clist;
+	int num_sinks = 0;
+	int i;
+
+	// align sink streams with their respective configurations
+	list_for_item(clist, &dev->bsink_list) {
+		sink = container_of(clist, struct comp_buffer, source_list);
+		if (sink->sink->state == dev->state) {
+			// if no config is set, then assign the sinks in order
+			if (!config) {
+				sinks[num_sinks++] = sink;
+				continue;
+			}
+
+			i = get_stream_index(config, sink->pipeline_id);
+			if (i < 0) {
+				comp_warn(dev, "crossover_assign_sinks(), could not assign sink %i",
+					  sink->pipeline_id);
+				continue;
+			}
+
+			if (sinks[i]) {
+				comp_warn(dev, "crossover_assign_sinks(), multiple sinks from pipeline %i are assigned",
+					  sink->pipeline_id);
+				continue;
+			}
+
+			sinks[i] = sink;
+			num_sinks++;
+		}
+	}
+
+	return num_sinks;
+}
+
+/**
+ * Sets the state of a single LR4 filter.
+ * An LR4 filter is built by cascading two biquads in series.
+ */
+static int crossover_init_coef_lr4(struct sof_crossover_config_lr4 *coef,
+				   lr4_state *lr4)
+{
+	/* Reuse the same coefficients for biquads so we only
+	 * store one copy of it. The processing functions will
+	 * feed the same coef array to both biquads.
+	 */
+	lr4->coef = (int32_t *)coef;
+
+	/* LR4 filters are two 2nd order filters, so only need 4 delay slots
+	 * delay[0..1] -> state for first biquad
+	 * delay[2..3] -> state for second biquad
+	 */
+	lr4->delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			     sizeof(uint64_t) * 4);
+	if (!lr4->delay)
+		return -ENOMEM;
+	return 0;
+}
+
+static void crossover_init_coef_ch(struct sof_crossover_config_lr4 *coef,
+				   struct crossover_state *state,
+				   int32_t num_sinks)
+{
+	int32_t i;
+	int32_t j = 0;
+
+	for (i = 0; i < num_sinks; i++) {
+		/* Get the low pass coefficients */
+		crossover_init_coef_lr4(&coef[j],
+					&state->lowpass[i]);
+		/* Get the high pass coefficients */
+		crossover_init_coef_lr4(&coef[j + 1],
+					&state->highpass[i]);
+		j += 2;
+	}
+}
+
+static int crossover_init_coef(struct comp_data *cd, int nch)
```

Comment:
this function is a monster rn, do you think it should be split up into smaller chunks?